### DISCUSSION TITLE
Problem with the AutoLeveler plugin (A java.lang.NullPointerException exception has occurred.)

### DISCUSSION THREAD

**Discussion by Filipdominik** [winder/Universal-G-Code-Sender#2345] [Help] [answered]
I got myself a CNC 3018 Pro, and want to mill some PCB's. I'm not expecting much just something usable. And to achieve this I need to level the surface in software as my copper boards are bend from the factory.
Anyways, when I set the machine on the start position to start milling, I want to run the AutoLeveler with these settings:
![image](https://github.com/winder/Universal-G-Code-Sender/assets/57330830/27ba7a33-9e6b-42b3-a2a1-9eb327df99f7)
![image](https://github.com/winder/Universal-G-Code-Sender/assets/57330830/e5aedd6c-d2f2-4f5a-aedd-21c9786ce850)

At this point I attempt to press the 'Scan Surface' button, but a java.lang.NullPointerException error gets thrown and no commands get send and the cnc doesn't move.  The full error log is listed at the end of this message.

What I've tried so far:

- (Re)Installing different version of UGS.
- Reinstalling Java to the most recent version
- Updating my Windows 10
- Restarting my computer
- Restarting the CNC controller (grbl v0.9j)

<i>java.lang.NullPointerException</i>
<i>&emsp;at com.willwinder.ugs.platform.surfacescanner.SurfaceScanner.scan(SurfaceScanner.java:190)</i>
<i>&emsp;at com.willwinder.ugs.platform.surfacescanner.actions.ScanSurfaceAction.actionPerformed(ScanSurfaceAction.java:67)</i>
<i>&emsp;at java.desktop/javax.swing.AbstractButton.fireActionPerformed(Unknown Source)</i>
<i>&emsp;at java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(Unknown Source)</i>
<i>&emsp;at java.desktop/javax.swing.DefaultButtonModel.fireActionPerformed(Unknown Source)</i>
<i>&emsp;at java.desktop/javax.swing.DefaultButtonModel.setPressed(Unknown Source)</i>
<i>&emsp;at java.desktop/javax.swing.plaf.basic.BasicButtonListener.mouseReleased(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.AWTEventMulticaster.mouseReleased(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.Component.processMouseEvent(Unknown Source)</i>
<i>&emsp;at java.desktop/javax.swing.JComponent.processMouseEvent(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.Component.processEvent(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.Container.processEvent(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.Component.dispatchEventImpl(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.Container.dispatchEventImpl(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.Component.dispatchEvent(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.Container.dispatchEventImpl(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.Window.dispatchEventImpl(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.Component.dispatchEvent(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.EventQueue.dispatchEventImpl(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.EventQueue$4.run(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.EventQueue$4.run(Unknown Source)</i>
<i>&emsp;at java.base/java.security.AccessController.doPrivileged(Unknown Source)</i>
<i>&emsp;at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)</i>
<i>&emsp;at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.EventQueue$5.run(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.EventQueue$5.run(Unknown Source)</i>
<i>&emsp;at java.base/java.security.AccessController.doPrivileged(Unknown Source)</i>
<i>&emsp;at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.EventQueue.dispatchEvent(Unknown Source)</i>
<i>[catch] at org.netbeans.core.TimableEventQueue.dispatchEvent(TimableEventQueue.java:136)</i>
<i>&emsp;at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.EventDispatchThread.pumpEvents(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.EventDispatchThread.pumpEvents(Unknown Source)</i>
<i>&emsp;at java.desktop/java.awt.EventDispatchThread.run(Unknown Source)</i>

Full log:
[UGS Autoleveler issue.txt](https://github.com/winder/Universal-G-Code-Sender/files/13032843/UGS.Autoleveler.issue.txt)

Any help or suggestions would be appreciated, maybe some other software that can achieve this too?

**Comment by breiler** [comment]
It seems that the work position has not been retrieved from the controller.

Can you post your settings `$$`?

Also, if you activate verbose logging in the console and copy about 20 lines of status reports when the machine is idle.

**Reply by Filipdominik** [reply]
From $$:
<i>>>$$
$0 = 10    (step pulse, usec)
$1 = 25    (step idle delay, msec)
$2 = 0    (step port invert mask:00000000)
$3 = 1    (dir port invert mask:00000001)
$4 = 0    (step enable invert, bool)
$5 = 0    (limit pins invert, bool)
$6 = 0    (probe pin invert, bool)
$10 = 1    (status report mask:00000001)
$11 = 0.010    (junction deviation, mm)
$12 = 0.002    (arc tolerance, mm)
$13 = 0    (report inches, bool)
$20 = 0    (soft limits, bool)
$21 = 0    (hard limits, bool)
$22 = 0    (homing cycle, bool)
$23 = 0    (homing dir invert mask:00000000)
$24 = 25.000    (homing feed, mm/min)
$25 = 500.000    (homing seek, mm/min)
$26 = 250    (homing debounce, msec)
$27 = 1.000    (homing pull-off, mm)
$100 = 402.208    (x, step/mm)
$101 = 401.606    (y, step/mm)
$102 = 401.633    (z, step/mm)
$110 = 1500.000    (x max rate, mm/min)
$111 = 1500.000    (y max rate, mm/min)
$112 = 1500.000    (z max rate, mm/min)
$120 = 10.000    (x accel, mm/sec^2)
$121 = 10.000    (y accel, mm/sec^2)
$122 = 10.000    (z accel, mm/sec^2)
$130 = 200.000    (x max travel, mm)
$131 = 200.000    (y max travel, mm)
$132 = 200.000    (z max travel, mm)
ok</i>
Note: the only settings I recall that I've changed are $100, $101 and $102 to calibrate the steps. Everything else should be default.


From verbose logging (right click on Console -> Enable Verbose Logging)
<i>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
<Idle,MPos:0.000,0.000,1.999>
</i>

**Reply by breiler** [reply, marked as answer]
Your settings for status report has likely been changed where work position is disabled from the report.

https://github.com/grbl/grbl/wiki/Configuring-Grbl-v0.9#10---status-report-maskbinary

Set $10=3 and it should work better.
